### ISSUE TITLE
[bug] Pop doesn't remove value if w.Write happened before sm.Pop

### ISSUE THREAD

**Issue by jackielii** [alexedwards/scs#215]
This is caused by https://github.com/alexedwards/scs/blob/7e11d57e8885fd38c604d4f79f0eef355650b997/session.go#L158-L160

TODO

**Comment by jum** [comment]
In general, all HTTP headers must be set and finalized before any content is written. This means you basically have to buffer the output if you intent to modify the session while producing output (very common when popping flash messages). This has bitten me quite a few times.

**Comment by jackielii** [comment]
> In general, all HTTP headers must be set and finalized before any content is written. This means you basically have to buffer the output if you intent to modify the session while producing output (very common when popping flash messages). This has bitten me quite a few times.

Yes, I spend one afternoon tracking down the problem. My call stack was too deep, I didn't even suspect session manager at first.

I think even if it's decided that this is expected, I prefer to have a error or even panic for this so that I know what not to do.

**Comment by alexedwards** [comment]
The expiry time issue described in point 5 has been fixed in 481af0a28857c8d0fbb2709d9623b03e7f67848b. The uncommitted modifications check and panic described in point 4 has been implemented in 2060319a03b0788d7b4562a865c57b42061c10bc. The documentation improvements described in points 2 and 3 have been made in 209de6e426de9259665975ce16b91331d228f052.

I'm going to close this issue as completed, but if you disagree then please let me know.
